ts/src/header-validator/validate-json.ts

```diff
@@ -285,61 +277,66 @@ function suitableScope(
 }
 
 function suitableOrigin(
-  ctx: Context,
   j: Json,
+  ctx: Context,
   rejectExtraComponents: boolean = false
 ): Maybe<string> {
-  return string(ctx, j).map((s) =>
-    suitableScope(ctx, s, 'origin', (u) => u.origin, rejectExtraComponents)
+  return string(j, ctx).map(
+    suitableScope,
```

Comment:
(general comment) I'm not 100% convinced that the approach is an improvement.

It looks good when you know what's going on.

But it is quite opaque at first sight.

Have you considered going the other way around meaning that we'd have the function receive their extra properties and return a function that receives JSON and the context object, partially applying properties when needed.

```ts
return string(j, ctx).map(
    suitableScope('origin', (u) => u.origin, rejectExtraComponents)
)
  ```

Reply:
I did think about that, but it means that every function that wants to be used this way needs to be return a function (effectively currying itself), whereas this approach works with the existing lambda approach *and* this binding shorthand.

If you think it's too hard, that's fine, and I can abandon.

Reply:
good with me to proceed with the suggested approach.